### PR TITLE
copr: fix unsound unsafe transmute

### PR DESCRIPTION
**Pull request by skyzh**
Signed-off-by: Alex Chi <iskyzh@gmail.com>

### What problem does this PR solve?

Problem Summary:

### What is changed and how it works?

As pointed out by @sticnarf https://github.com/tikv/tikv/pull/9648#pullrequestreview-643187256 , transmute of enum value is unsound. This PR intends to fix this.

<del>For the first step, we just let the compiler fail. I'll propose a fix soon.</del>

Now we have made all transmute sound by retaining lifetime `retain_lifetime_transmute`.

What's Changed:

### Check List 

Tests 

- Unit test

Side effects

### Release note 
* No release note

### CODE REVIEW

File: components/tidb_query_datatype/src/codec/mysql/enums.rs [hunk 1]
```diff
@@ -186,14 +190,14 @@ impl<'a> ToString for EnumRef<'a> {
 pub trait EnumEncoder: NumberEncoder {
     #[inline]
     fn write_enum(&mut self, data: EnumRef) -> Result<()> {
-        self.write_u64_le(data.value as u64)?;
+        self.write_u64_le(*data.value as u64)?;
```
Comment (dyzsr) [inline components/tidb_query_datatype/src/codec/mysql/enums.rs:193]: I forgot why I wrote `as u64` here😂. I think we can remove them.

Comment (skyzh) [reply components/tidb_query_datatype/src/codec/mysql/enums.rs]: I've removed that :)

**Comment by dyzsr** [comment]
LGTM

**Review by sticnarf — APPROVED** [review]
LGTM

**Comment by Xuanwo** [comment]
LGTM

**Review by Xuanwo — APPROVED** [review]

**Pull request merged**